File: app/src/main/java/com/pr0gramm/app/services/preloading/DatabasePreloadManager.java
package com.pr0gramm.app.services.preloading;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Build;

import com.google.common.base.Optional;
import com.google.common.collect.ImmutableCollection;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import com.pr0gramm.app.util.AndroidUtility;
import com.pr0gramm.app.util.BackgroundScheduler;
import com.squareup.sqlbrite.BriteDatabase;

import org.joda.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;

/**
 */
@Singleton
public class DatabasePreloadManager implements PreloadManager {
    private static final Logger logger = LoggerFactory.getLogger("DatabasePreloadManager");

    private static final String TABLE_NAME = "preload_2";
    private static final String QUERY_ALL_ITEM_IDS = "SELECT * FROM " + TABLE_NAME;

    private final Observable<BriteDatabase> database;
    private final AtomicReference<ImmutableMap<Long, PreloadItem>> preloadCache =
            new AtomicReference<>(ImmutableMap.<Long, PreloadItem>of());

    @Inject
    public DatabasePreloadManager(Observable<BriteDatabase> database) {
        this.database = database;

        // initialize in background.
        queryAllItems()
                .subscribeOn(BackgroundScheduler.instance())
                .subscribe(this::setPreloadCache);
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    private void setPreloadCache(ImmutableMap<Long, PreloadItem> items) {
        List<PreloadItem> missing = new ArrayList<>();
        ImmutableMap.Builder<Long, PreloadItem> available = ImmutableMap.builder();

        for (PreloadItem item : items.values()) {
            if (!item.thumbnail().exists() || !item.media().exists()) {
                missing.add(item);
            } else {
                available.put(item.itemId(), item);
            }
        }

        if (missing.size() > 0) {
            // delete missing entries in background.
            this.database.subscribeOn(BackgroundScheduler.instance()).subscribe(db -> {
                try (BriteDatabase.Transaction tx = db.newTransaction()) {
                    deleteTx(db, missing);
                    tx.markSuccessful();
                }
            });
        }

        preloadCache.set(available.build());
    }

    private Observable<ImmutableMap<Long, PreloadItem>> queryAllItems() {
        return this.database
                .flatMap(db -> db.createQuery(TABLE_NAME, QUERY_ALL_ITEM_IDS).mapToList(this::readPreloadItem))
                .map(this::readPreloadEntriesFromCursor)
                .doOnError(AndroidUtility::logToCrashlytics)
                .onErrorResumeNext(Observable.empty());
    }

    private PreloadItem readPreloadItem(Cursor cursor) {
        int cItemId = cursor.getColumnIndexOrThrow("itemId");
        int cCreation = cursor.getColumnIndexOrThrow("creation");
        int cMedia = cursor.getColumnIndexOrThrow("media");
        int cThumbnail = cursor.getColumnIndexOrThrow("thumbnail");

        return ImmutablePreloadItem.builder()
                .itemId(cursor.getLong(cItemId))
                .creation(new Instant(cursor.getLong(cCreation)))
                .media(new File(cursor.getString(cMedia)))
                .thumbnail(new File(cursor.getString(cThumbnail)))
                .build();
    }

    private ImmutableMap<Long, PreloadItem> readPreloadEntriesFromCursor(List<PreloadItem> items) {
        return Maps.uniqueIndex(items, PreloadItem::itemId);
    }

    /**
     * Inserts the given entry blockingly into the database.
     */
    @Override
    public void store(PreloadItem entry) {
        ContentValues values = new ContentValues();
        values.put("itemId", entry.itemId());
        values.put("creation", entry.creation().getMillis());
        values.put("media", entry.media().getPath());
        values.put("thumbnail", entry.thumbnail().getPath());
        db().insert(TABLE_NAME, values, SQLiteDatabase.CONFLICT_REPLACE);
    }

    /**
     * Checks if an entry with the given itemId already exists in the database.
     */
    @Override
    public boolean exists(long itemId) {
        return preloadCache.get().containsKey(itemId);
    }

    /**
     * Returns the {@link PreloadItem} with a given id.
     */
    @Override
    public Optional<PreloadItem> get(long itemId) {
        return Optional.fromNullable(preloadCache.get().get(itemId));
    }

    @SuppressLint("NewApi")
    @Override
    public void deleteBefore(Instant threshold) {
        logger.info("Removing all files preloaded before {}", threshold);

        BriteDatabase db = db();

        try (BriteDatabase.Transaction tx = db.newTransaction()) {
            List<PreloadItem> items = new ArrayList<>();
            try (Cursor cursor = db.query("SELECT * FROM " + TABLE_NAME + " WHERE creation<?",
                    String.valueOf(threshold.getMillis()))) {

                while (cursor.moveToNext())
                    items.add(readPreloadItem(cursor));
            }

            deleteTx(db, items);
            tx.markSuccessful();
        }
    }

    private void deleteTx(BriteDatabase db, Iterable<PreloadItem> items) {
        for (PreloadItem item : items) {
            logger.info("Removing files for itemId={}", item.itemId());

            if (!item.media().delete())
                logger.warn("Could not delete media file {}", item.media());

            if (!item.thumbnail().delete())
                logger.warn("Could not delete thumbnail file {}", item.thumbnail());

            // delete entry from database
            db.delete(TABLE_NAME, "itemId=?", String.valueOf(item.itemId()));
        }
    }

    /**
     * Returns a list of all preloaded items.
     */
    public Observable<ImmutableCollection<PreloadItem>> all() {
        return queryAllItems().map(ImmutableMap::values);
    }

    private BriteDatabase db() {
        return database.toBlocking().single();
    }

    public static void onCreate(SQLiteDatabase db) {
        logger.info("initializing sqlite database");
        db.execSQL("CREATE TABLE IF NOT EXISTS " + TABLE_NAME + " (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "itemId INT NOT NULL UNIQUE," +
                "creation INT NOT NULL," +
                "media TEXT NOT NULL," +
                "thumbnail TEXT NOT NULL)");
    }
}

File: app/src/main/java/com/pr0gramm/app/services/preloading/DatabasePreloadManager.kt
package com.pr0gramm.app.services.preloading

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.pr0gramm.app.Instant
import com.pr0gramm.app.services.preloading.PreloadManager.PreloadItem
import com.pr0gramm.app.util.*
import com.pr0gramm.app.util.Databases.withTransaction
import com.squareup.sqlbrite.BriteDatabase
import gnu.trove.map.TLongObjectMap
import gnu.trove.map.hash.TLongObjectHashMap
import rx.Observable
import java.io.File
import java.util.*
import java.util.concurrent.atomic.AtomicReference


/**
 */

class DatabasePreloadManager(private val database: BriteDatabase) : PreloadManager {
    private val preloadCache = AtomicReference(
            TLongObjectHashMap<PreloadItem>() as TLongObjectMap<PreloadItem>)

    private val observeAllItems = Observable.just(database)
            .flatMap { db ->
                db.createQuery(TABLE_NAME, QUERY_ALL_ITEM_IDS)
                        .mapToList { readPreloadItem(it) }
                        .map { convertToMap(it) }
                        .onErrorResumeEmpty()
            }
            .share()

    init {
        // initialize and cache in background for polling
        observeAllItems.subscribe { setPreloadCache(it) }
    }

    private fun setPreloadCache(items: TLongObjectMap<PreloadItem>) {
        val missing = mutableListOf<PreloadItem>()
        val available = TLongObjectHashMap<PreloadItem>(items.size())

        for (item in items.valueCollection()) {
            if (!item.thumbnail.exists() || !item.media.exists()) {
                missing.add(item)
            } else {
                available.put(item.itemId, item)
            }
        }

        if (missing.isNotEmpty()) {
            // delete missing entries in background.
            doInBackground {
                withTransaction(database) {
                    deleteTx(database, missing)
                }
            }
        }

        preloadCache.set(available)
    }

    private fun readPreloadItem(cursor: Cursor): PreloadItem {
        val cItemId = cursor.getColumnIndexOrThrow("itemId")
        val cCreation = cursor.getColumnIndexOrThrow("creation")
        val cMedia = cursor.getColumnIndexOrThrow("media")
        val cThumbnail = cursor.getColumnIndexOrThrow("thumbnail")

        return PreloadItem(
                itemId = cursor.getLong(cItemId),
                creation = Instant(cursor.getLong(cCreation)),
                media = File(cursor.getString(cMedia)),
                thumbnail = File(cursor.getString(cThumbnail)))
    }

    private fun convertToMap(items: List<PreloadItem>): TLongObjectMap<PreloadItem> {
        val map = TLongObjectHashMap<PreloadItem>(items.size)
        items.forEach { item -> map.put(item.itemId, item) }
        return map
    }

    /**
     * Inserts the given entry blockingly into the database. Must not be run on the main thread.
     */
    override fun store(entry: PreloadItem) {
        checkNotMainThread()

        val values = ContentValues()
        values.put("itemId", entry.itemId)
        values.put("creation", entry.creation.millis)
        values.put("media", entry.media.path)
        values.put("thumbnail", entry.thumbnail.path)
        database.insert(TABLE_NAME, values, SQLiteDatabase.CONFLICT_REPLACE)
    }

    /**
     * Checks if an entry with the given itemId already exists in the database.
     */
    override fun exists(itemId: Long): Boolean {
        return preloadCache.get().containsKey(itemId)
    }

    /**
     * Returns the [PreloadItem] with a given id.
     */
    override operator fun get(itemId: Long): PreloadItem? {
        return preloadCache.get().get(itemId)
    }

    override fun deleteBefore(threshold: Instant) {
        logger.info { "Removing all files preloaded before $threshold" }

        withTransaction(database) {
            val items = ArrayList<PreloadItem>()
            database.query("SELECT * FROM $TABLE_NAME WHERE creation<?",
                    threshold.millis.toString()).use { cursor ->

                while (cursor.moveToNext())
                    items.add(readPreloadItem(cursor))
            }

            deleteTx(database, items)
        }
    }

    private fun deleteTx(db: BriteDatabase, items: Iterable<PreloadItem>) {
        for (item in items) {
            logger.info { "Removing files for itemId=${item.itemId}" }

            if (!item.media.delete())
                logger.warn { "Could not delete media file ${item.media}" }

            if (!item.thumbnail.delete())
                logger.warn { "Could not delete thumbnail file ${item.thumbnail}" }

            // delete entry from database
            db.delete(TABLE_NAME, "itemId=?", item.itemId.toString())
        }
    }

    /**
     * Returns a list of all preloaded items.
     */
    override fun all(): Observable<Collection<PreloadItem>> {
        return observeAllItems.map { it.valueCollection() }
    }

    companion object {
        private val logger = Logger("DatabasePreloadManager")

        private const val TABLE_NAME = "preload_2"
        private const val QUERY_ALL_ITEM_IDS = "SELECT * FROM $TABLE_NAME"

        fun onCreate(db: SQLiteDatabase) {
            logger.info { "initializing sqlite database" }
            db.execSQL("""CREATE TABLE IF NOT EXISTS $TABLE_NAME (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    itemId INT NOT NULL UNIQUE,
                    creation INT NOT NULL,
                    media TEXT NOT NULL,
                    thumbnail TEXT NOT NULL)""")
        }
    }
}
